kubectl: `apply --server-side` managed fields migration

alexzielenski opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR will transparently take ownership of any client-side-apply fields when `--server-side` apply is used. If this is not done, then it will not be possible to remove fields previously used by client-side-apply with server-side-apply (without manually patching managed fields yourself).

Previously this change was [proposed in a KEP](https://github.com/kubernetes/enhancements/pull/3518), but during discussion with the SIG we have [scoped it down to a bug fix](https://github.com/kubernetes/enhancements/pull/3518#discussion_r984724392).

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

addresses issues brought up in the following and likely more bug reports

Fixes #107980
Fixes #108081
Fixes #107417
Fixes #112826
Fixed #99003

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
For `kubectl`, `--server-side` now migrates ownership of all fields used by client-side-apply to the specified `--fieldmanager`. This prevents fields previously specified using kubectl from being able to live outside of server-side-apply's management and become undeleteable.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [Discussion]: https://github.com/kubernetes/enhancements/pull/3518#discussion_r984724392
```

/sig cli
/cc @soltysh @KnVerey @apelisse

alexzielenski commented:
/hold still doing some manual testing, but sharing this PR to show general approach

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +543,53 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		//
+		// To check beforehand means 2 requests for most operations, and 3
+		//	requests in worst case.
+		if patchData, err := o.calculateSSAMigrationPatch(obj); err != nil {
```
This approach differs a big from the approach discussed in the SIG. During the SIG, we proposed for SSA to unconditionally fetch the object and patch the fields BEFORE the apply is done.

I propose this alternative method which preserve's SSA's property of only needing 1 request in the average case. But it does complicate the error handling quite a bit.

Thoughts?

seans3 replied:
This is a better approach. No extra retrievals for the common (no migration) case is an excellent approach.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:
```diff
@@ -710,6 +713,82 @@ func TestApplyPruneObjects(t *testing.T) {
 
+// Tests that apply of object in need of CSA migration results in a call
+// to patch it.
+func TestApplyCSAMigration(t *testing.T) {
```
This is a very basic unit test. I am unsure how kubectl testing is organized completely. 

I'd like to add a test that shows a client-side-apply followed by a server-side-apply does not result in undeleteable fields. Need an e2e test for that I assume

ardaguclu replied:
integration tests would be very useful for this fix.

apelisse replied:
There's also a series of tests available in https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-cmd.sh, which is a little messy and hard to debug, but can be used to test what you want I think.

seans3 replied:
We'll need to add significant unit tests. Integration tests (including `make test-cmd`) are helpful, but the focus must be on the unit tests.

alexzielenski commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -50,10 +50,13 @@ var csaAnnotationFieldSet = fieldpath.NewSet(fieldpath.MakePathOrDie("metadata",
 	obj runtime.Object,
 	csaManagerName string,
 	ssaManagerName string,
+	keepLastApplied bool,
```
As discussed in SIG, we would like to retain the last applied annotation for now. This is OK, since server-side-apply has logic to keep it up to date from the server-side if it exists.

apelisse replied:
This methods isn't used yet, it's still brand new, but we have people using server-side apply outside the context of kubectl, and they may want to get rid of the last applied annotation.

That being said, I think if someone wants to remove the lastApplied annotation, it'd be simpler to have a `removeLastApplied()` (pretty sure it already exists in some form), rather than move the burden of specifying that parameter to everyone.

alexzielenski reviewed (commented):
left some comments/discussion points

ardaguclu reviewed (commented):
I'm very happy to see that this known problem is being fixed.

sftim commented:
For the release note, can I suggest:
```markdown
For `kubectl`, `--server-side` now migrates ownership of all fields used by client-side-apply to the specified `--fieldmanager`. This prevents fields previously specified using kubectl from being able to live outside of server-side-apply's management and become undeleteable.
```
?

fedebongio commented:
/triage accepted

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +543,53 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+			)
+
+			if err != nil {
+				// Print as warning? Non-fatal since patch was already applied
```
If the patch failed, then it means that CSA will continue to own the fields, and they will not be removed.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +710,47 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+	upgraded := obj.DeepCopyObject()
+	if err := csaupgrade.UpgradeManagedFields(
+		upgraded,
+		"kubectl-client-side-apply",
```
This hard-code is really unfortunate, on the other hand, I don't want people to specify two manager when they run kubectl...

seans3 replied:
Unfortunate, but not a blocker. Is this enumerated as a constant somewhere?

alexzielenski replied:
yes, there was `FieldManagerClientSideApply` in this very file. Will replace

seans3 commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +543,53 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+			// (non-fatal because object was successfully applied, but it might
+			// have issues since migration failed?)
+		} else if patchData != nil {
+			// If patchData is non-nil there is no patch necessary
```
Not sure I understand this sentence. I would have written

`// Update object's managed fields with CSA managed fields merged into SSA manager.`

apelisse replied:
Good catch, I forgot to comment, this one was definitely wrong, thanks Sean :-)

seans3 commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +710,47 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 	return nil
 }
 
+func (o *ApplyOptions) calculateSSAMigrationPatch(obj runtime.Object) ([]byte, error) {
```
While this function is not exported, I would add at least some comments describing the patch created. And this function should have strong unit test coverage.

seans3 replied:
While it's not common to unit test helper functions, in the case where they are particularly intricate or dense, we should. And this function has straightforward inputs/outputs that make unit testing easier. Setting up the fake `ApplyOptions` for the unit tests is accomplished in several other unit tests, including in the `cli-utils` repository: https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/apply/task/apply_task.go#L188

alexzielenski replied:
Is this function super dense or intricate? I'm sure sure unit testing it specifically has much value especially given the `UpgradeManagedFields` is already well tested. 

The only additional logic unit tests for this function would add coverage for is the `jsonPatch` construction - and I'm imagining that test would basically just have this same code?

seans3 commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +543,53 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		// Check if the returned object needs to have its
+		// kubectl-client-side-apply managed fields migrated server-side-apply.
+		//
+		// By checking afterward instead of fetching the object beforehand and
```
Avoiding extra retrievals in the common case (no migration) is an excellent approach. And two extra REST calls for the uncommon migration case seems like a very palatable price to pay.

seans3 commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade_test.go:
```diff
@@ -462,6 +462,7 @@ metadata:
 				upgraded,
 				testCase.CSAManager,
 				testCase.SSAManager,
+				false,
```
Agree with @ardaguclu

seans3 reviewed (commented):
This PR is in a good place. This PR has a lot of useful comments -- we probably need more. It would be useful to describe the migration process high level somewhere. Something like: 
```
field ownership metadata is stored in three places: server-side-manager
fields, client-side-manager fields, and the last_applied_configuration annotation.
The migration merges the client-side-manager fields into the server-side-manager
fields, and removes the last_applied_annotation (if keepLastApplied is false), so
the only remaining metadata is server-side-manager fields.
```
Warnings will be fine in the places you've enumerated. There are examples in the `apply.go` file showing how to print warnings.

The one area that we'll need to work on is the tests. Current `kubectl apply` unit tests are dense, but we'll need to add significant unit tests to ensure proper coverage. As a general rule, we should not have lower unit test coverage numbers after a PR. Including before/after unit test coverage numbers in the PR would be useful. Integration tests will be helpful, but the
focus should be on unit tests. Let's get together to see how to test this.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +543,53 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		// To check beforehand means 2 requests for most operations, and 3
+		//	requests in worst case.
+		if patchData, err := o.calculateSSAMigrationPatch(obj); err != nil {
+			// Print-error as a warning? Not sure how to treat this non-fatal error
```
warning added

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +722,46 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 	return nil
 }
 
+func (o *ApplyOptions) calculateSSAMigrationPatch(obj runtime.Object) ([]byte, error) {
```
I would move this whole function into client-go for easier consumption by other actors that want to migrate.

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade_test.go:
```diff
@@ -440,11 +458,12 @@ metadata:
     time: "2022-08-23T23:08:23Z"
   name: test
-  namespace: default  
+  namespace: default
```
The whole client-go part lgtm.

apelisse replied:
Just missing the calculateCSAPatch IMO.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +544,64 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		// kubectl-client-side-apply managed fields migrated server-side-apply.
+		//
+		// field ownership metadata is stored in three places:
+		// 	- server-side managed fields
+		//  - client-side managed fields
```
is gofmt aligning things like this?!

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +544,64 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+			} else {
+				// Re-send original SSA patch (this will allow dropped fields to
+				// finally be removed)
+				obj, err = helper.Patch(
```
Just to clarify: since `obj` is updated here, and later "Refreshed" in the object, I think the result of `kubectl apply --server-side -o yaml` will show the latest applied object, which is certainly what we want.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -163,6 +166,7 @@ var (
 
 	warningNoLastAppliedConfigAnnotation = "Warning: resource %[1]s is missing the %[2]s annotation which is required by %[3]s apply. %[3]s apply should only be used on resources created declaratively by either %[3]s create --save-config or %[3]s apply. The missing annotation will be patched automatically.\n"
 	warningChangesOnDeletingResource     = "Warning: Detected changes to resource %[1]s which is currently being deleted.\n"
+	warningMigrationPatchFailed          = "Warning: attempted migration of client-side-apply managed fields to server-side-apply failed. Please re-apply to try again. Error: %[1]s\n"
```
We're definitely going to have people wonder what's going on when they read that error. I'm trying to see what I would think and how would I debug given such an error:
1. Have I requested that migration? Why does it think it should migrate?
2. What is it migrating and how?

That's even more true since most of this migration is done behind-the-scene without letting users know what's going on.  Is there anything we can do to explain what's going on?

apelisse replied:
I would make the warnings a little more contextual (based on what's actually going on), they show the error but since the same string is used in multiple location, it might be hard to investigate, even for us.

apelisse reviewed (commented):
Mostly good, thanks, missing test-cmd as discussed offline.

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -107,21 +134,76 @@ func UpgradeManagedFields(
+	operation metav1.ManagedFieldsOperationType,
+	managerName string,
+	subresource string,
+	paths *fieldpath.Set,
```
Will the paths depend on the apiVersion? They probably should. Since you're already passing this as a param, that's nice, but I would put them together since they are coupled.

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -107,21 +134,76 @@ func UpgradeManagedFields(
-	// Commit changes to object
-	accessor.SetAnnotations(annotations)
-	accessor.SetManagedFields(filteredManagers)
+// Ensures a given manager exists that owns the given field paths
+// If there is an entry for the given manager already, inserts
+//
+//	field path into its ownership set.
+//
+// Otherwise, creates manager at latest APIVersion which owns the field
+// Returns an error if there was an issue with decoding/encoding. Or nil if
+// the given field path are now owned by the desired manager
```
I'd really rather avoid doing that if we can, playing with managed fields like this is getting more and more tedious :-/

apelisse replied:
We should also probably unit-test this function if we end-up keeping it, but let's wait for now.

apelisse reviewed (commented):
I haven't relooked at all the details again since we're not super sure  yet, but I've looked at it enough times that I know this will be close when we agree on how to do it, thanks Alex!

seans3 commented:
/priority important-soon

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -88,16 +148,16 @@ func UpgradeManagedFields(
 			// There are no CSA managers that need to be converted. Nothing to do
 			// Return early
-			return nil
+			return nil, nil
```
so if there's nothing to do, you return nil (this function is supposed to return the new `ManagedFieldsEntry`) as if the managed fields needed to be dropped? And in `UpgradeManagedFieldsPatch` you set that directly into the patch, so would you sometime create a patch that deletes managedFields?

alexzielenski replied:
I appreciate you writing this eagle-eyed comment in the form of a question xD. I'll have to add a new test case and handle this case correctly in `UpgradeManagedFieldsPatch`

alexzielenski replied:
fixed

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -492,6 +496,11 @@ func (o *ApplyOptions) applyOneObject(info *resource.Info) error {
+	lastAppliedHelper := resource.NewHelper(info.Client, info.Mapping).
+		DryRun(o.DryRunStrategy == cmdutil.DryRunServer).
+		WithFieldManager(FieldManagerClientSideApply).
+		WithFieldValidation(o.ValidationDirective)
```
> it unconditionally replaces the field manager.

What do you mean by that?

alexzielenski replied:
Regardless what field manager I put into `PatchOptions`, the helper overrides it:

https://github.com/kubernetes/kubernetes/blob/17dd76f5d49848e0bcb057035cda7946b578584f/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go#L256-L258

alexzielenski replied:
right now I am doing something dirty

https://github.com/kubernetes/kubernetes/blob/bce39b8b7897097cca8309c2a8653772248d3aa1/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go#L703-L710

alexzielenski replied:
new changes include refactor to the helper which respects the field manager on the options for the patch

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -492,6 +496,11 @@ func (o *ApplyOptions) applyOneObject(info *resource.Info) error {
 
+	lastAppliedHelper := resource.NewHelper(info.Client, info.Mapping).
+		DryRun(o.DryRunStrategy == cmdutil.DryRunServer).
+		WithFieldManager(FieldManagerClientSideApply).
```
I haven't read the whole thing yes, but that sounds like the manager you'll use to apply the last applied annotation, didn't we agree to use a different field manager for that?

alexzielenski replied:
It is a different field manager since its operation is `Apply` instead of `Update` :). Happy to change if that isnt convincing enough

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +687,132 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+
+	if err != nil {
+		// Print-error as a warning.
+		// This is a non-fatal error because object was successfully applied,
+		// but it might have issues since migration failed.
+		//
+		// This migration will be re-attempted if neccessary upon next
+		// apply.
+		fmt.Fprintf(o.ErrOut, warningMigrationPatchGenFailed, err.Error())
+		return obj, err
+	} else if patchData == nil {
+		// nil patch data means nothing to do - object is already migrated
+		return obj, nil
+	}
+
+	// Send the patch to upgrade the managed fields if it is non-nil
+	_, err = helper.Patch(
+		namespace,
+		name,
+		types.JSONPatchType,
+		patchData,
+		nil,
+	)
+
+	if err != nil {
+		// Will get StatusUnprocessableEntity our RV test failed due to a
+		// race condition
+		fmt.Fprintf(o.ErrOut, warningMigrationPatchFailed, err.Error())
+		return obj, nil
+	}
+
+	// Re-send original SSA patch (this will allow dropped fields to
+	// finally be removed)
+	return helper.Patch(
+		namespace,
+		name,
+		types.ApplyPatchType,
+		data,
+		&options,
+	)
```
Im assuming you are referring to the `JSONPatch` and not the `ApplyPatch` to which the comment is attributed?

alexzielenski commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,58 @@ run_kubectl_server_side_apply_tests() {
+EOF
+  )
+
+  kube::test::if_has_string "${output_message}" 'configmap/test serverside-applied'
```
added to test

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +695,134 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+	}
+
+	// This operastion is already completed. Nothing to do
+	if lastAppliedSeparated {
+		return obj, nil
+	}
```
you can still remove this variable and return from within the loop, no?

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +695,134 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 	return nil
 }
 
+// Saves the last-applied-configuration annotation in a separate SSA field manager
+// to prevent it from being dropped by users who have transitioned to SSA.
+//
+// If this operation is not performed, then the last-applied-configuration annotation
+// would be removed from the object upon the first SSA usage.
```
We only want to persist the annotation if it was already present right? I think as it reads now, the annotation will be added for all server-side applied objects.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +695,134 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		types.ApplyPatchType,
+		modified,
+		&metav1.PatchOptions{
+			FieldManager: "kubectl-last-applied",
```
Maybe use a constant for this (I like its name), and also that's not the name you're looking for at the beginning of the function?

alexzielenski replied:
yeah this field manager value is ignored. I was trying to explain it gets overridden by the helper

apelisse commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,87 @@ run_kubectl_server_side_apply_tests() {
+metadata:
+  name: test
+data:
+  key: value
```
If you client-side apply without this key, does it get removed?

alexzielenski replied:
I added a component to the integration test where client-side-apply removes a key added by server-side-apply.

KnVerey commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,87 @@ run_kubectl_server_side_apply_tests() {
+  kube::test::get_object_assert "configmap test" "{{ .data.legacy }}" '<no value>'
+
+  # CSA the object after it has been server-side-applied and had a field removed
+  # Add new key with client-side-apply.
```
Could we have this part of the test attempt a deletion as well, to prove that the downgrade is working properly? (or do we already have one in the suite that would fail without the annotation retention code?)

alexzielenski replied:
Am I understanding correctly that you are requesting here to remove  a field in the test, similar to https://github.com/kubernetes/kubernetes/pull/112905#discussion_r1010993799 ?

KnVerey replied:
Yeah, it's the same request. I was doing my review in a tab I hadn't refreshed in hours. 🤦‍♀️

alexzielenski replied:
copying comment here too:

I added a component to the integration test where client-side-apply removes a key added by server-side-apply.

KnVerey commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,87 @@ run_kubectl_server_side_apply_tests() {
   # clean-up
   kubectl delete -f hack/testdata/pod.yaml "${kube_flags[@]:?}"
 
+    # Test apply migration
```
[#107980](https://github.com/kubernetes/kubernetes/issues/107980), one of the issues cited as fixed, seems to be about kubectl diff. We're not specifically modifying that command, though I can imagine how the fix to the persisted state might be sufficient. Have you confirmed that specifically? If it is fixed, should we add a regression test for that too?

alexzielenski replied:
Yes I've also added an integration test to show that after doing the migration to server-side-apply, an object which previously would show a conflict upon a server-side diff now correctly shows the diff.

KnVerey reviewed (commented):
I still need to look more deeply in some places, but this is looking good to me. 👍 
I think we can add the heavily commented https://github.com/kubernetes/kubernetes/issues/99003 to the list of fixed issues too?

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -540,6 +545,36 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		} else if obj, err = o.migrateToSSAIfNecessary(helper, obj); err != nil {
+			// Print-error as a warning.
+			// This is a non-fatal error because object was successfully applied
+			// above, but it might have issues since migration failed.
```
Thanks for digging up that precedent Katrina. I think what you suggest is a good plan:

1. retry a few times in a loop if there was a conflict
2. fail if we could not succeed after a number of attempts

alexzielenski replied:
thinking about this some more. the context on the server is a bitt different from here. The server is making that check before committing the object to etcd. Here the patch has already succeeded. For that reason I'm thinking maybe keep the warning

KnVerey replied:
Ah ok, that makes sense and I'm fine with the warning.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +695,134 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+	newObj, err := helper.Patch(
+		accessor.GetNamespace(),
+		accessor.GetName(),
+		types.ApplyPatchType,
```
correct

alexzielenski commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -107,21 +175,17 @@ func UpgradeManagedFields(
 
-	// Wipe out last-applied-configuration annotation if it exists
-	annotations := accessor.GetAnnotations()
-	delete(annotations, csaAnnotationName)
```
They would be independent CSA implementations which likely do not care about the annotation.

I agree with the sentiment that is is a cleaner separation to remove the logic from client-go!

apelisse commented on staging/src/k8s.io/cli-runtime/pkg/resource/helper.go:
```diff
@@ -273,13 +273,13 @@ func (m *Helper) Patch(namespace, name string, pt types.PatchType, data []byte,
 		options.FieldManager = m.FieldManager
 	}
-	if m.FieldValidation != "" {
+	if m.FieldValidation != "" && options.FieldValidation == "" {
```
This is a lot of copy-paste, all in the same helper. Does the helper need some help?

alexzielenski replied:
That sure would be something but unfortunately the types are all different: `metav1.DeleteOptions`, `metav1.PatchOptions`, `metav1.UpdateOptions`. I'm not sure how to share the code for this in Go

apelisse replied:
sad face emoji

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -18,17 +18,44 @@ package csaupgrade
+		}
+
+		if fields.Difference(&fieldSet).Empty() {
+			result = append(result, entry.Manager)
```
Nit, you could have returned the `ManagedFieldsEntry`? And called the function `FilterFieldOwners`? Actually, there's possibly more than one field, so maybe `FindFieldsOwners` or `FilterFieldsOwners`

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade_test.go:
```diff
@@ -469,14 +562,50 @@ metadata:
+				if !reflect.DeepEqual(&patched, upgraded) {
+					t.Fatalf("expected patch to produce an upgraded object: %v", cmp.Diff(patched, upgraded))
+				}
+			}
```
Where's the test for `FindFieldOwners`? :-/

alexzielenski replied:
there now!

apelisse commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -18,17 +18,44 @@ package csaupgrade
+			continue
+		}
+
+		if fields.Difference(&fieldSet).Empty() {
```
I can't remember is this is symetric, is it symetric? I was hoping the tests would tell me ;-)

alexzielenski replied:
what do you mean by symmetric?

apelisse replied:
No, Ok, I think this works, but a test still would be nice

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +697,170 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+				return err
+			}
+
+			// Refresh the object for next iteration
+			err = info.Get()
+			if err != nil {
+				// If there was an error fetching, return error
+				return err
+			}
```
Do we need to do that? The returned object above should be good enough (it's basically the same as doing a get).

alexzielenski replied:
is there an object returned when there is an error?

apelisse replied:
I mis-read, hide this comment :-)

apelisse commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,160 @@ run_kubectl_server_side_apply_tests() {
+  kube::test::get_object_assert "configmap ssa-test" "{{ .data.legacy }}" '<no value>'
+
+  # Clean up
+  kubectl delete configmap ssa-test
```
We have more logic now, I haven't thought this through, but I'm sure we're missing some tests for these new cases?

alexzielenski replied:
I dont understand

apelisse replied:
I guess we know have additional logic to detect what is the name of the client-side manager, so maybe we should try that too? Client-side apply with a different manager, and see if it gets converted?

alexzielenski replied:
There is a test for that case already (that migration succeeds when using a custom manager). It is the test case this discussion is attached to. Is it not sufficient?

The test performs this sequence using the `myfm` field manager:

```shell
kubectl apply --server-side=false --field-manager=myfm -f - <a configmap>
kubectl apply --server-side=true --field-manager=myfm  -f - <the same object>
kubectl apply --server-side=true --field-manager=myfm  -f - <a slightly changed object>
```

tests that a key from csa can be dropped

apelisse replied:
OK Yeah, that's good then.

alexzielenski committed:
add FindFieldsOwners util function

to be used by kubectl to determine csa manager name used

findowners

alexzielenski committed:
remove kubectl annotation logic from upgrade patch

adds unneccessary complexity. also discussed in SIG CLI meeting to keep annotation around for a while longer

alexzielenski commented:
/retest

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +697,170 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+//
+// If this operation is not performed, then the last-applied-configuration annotation
+// would be removed from the object upon the first SSA usage.
+//
```
Is it worth mentioning _why_ we want to keep it even though the user is on SSA? I.e. that it is required for downgrade to be possible (per the [docs](https://kubernetes.io/docs/reference/using-api/server-side-apply/#downgrading-from-server-side-apply-to-client-side-apply) and the [KEP](https://github.com/kubernetes/kubernetes/pull/90187))

alexzielenski replied:
added doc links

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +697,170 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+	originalAnnotation, hasAnnotation := annots[corev1.LastAppliedConfigAnnotation]
+
+	// If the annotation does not already exist, we do not do anything
+	if !hasAnnotation {
```
For this one, I'm wondering if we should also skip if the _current_ field manager name isn't `kubectl`. The current behaviour always preserves the annotation because it continues to be owned by the CSA manager, BUT the downgrade doesn't actually work, because the annotation is not being kept in sync. So it would be a behaviour change... but arguably the retention when it is not kept correct is a bug. Wdyt?

alexzielenski replied:
To be clear, this is a suggestion to drop the annotation if during the migration the user is not using the default `kubectl` field manager? I am on board but want to be clear that this entails the annotation gets dropped if at the time of the first migration, the field manager is changed.

KnVerey replied:
> want to be clear that this entails the annotation gets dropped if the field manager is not kubectl at the time of the first migration.

Yes, that's what I'd expect. Can you think of any reason it would be valuable to retain in that case? The docs explicitly say that custom field managers opt out of the downgrade capability. Imagining the following sequence:
1. `kubectl apply -f mycm.yaml`
2. `kubectl apply -f mycm.yaml --server-side --field-manager=custom`
3. `kubectl apply -f mycm.yaml`

If step 2 made any changes, step 3 is going to use broken logic whether or not we keep the annotation, because the data it is using is either empty (behaves like a fresh apply) or stale (behaviour undefined).  I guess if they instead do:

1. `kubectl apply -f mycm.yaml`
2. `kubectl apply -f mycm.yaml --server-side --field-manager=custom`
3. `kubectl apply -f mycm.yaml --server-side --force-conflicts`
4. `kubectl apply -f mycm.yaml`

Then step 3 will repair the annotation if it is still present? I guess that could be considered useful, but I don't like the tradeoff of essentially undefined behaviour (instead of "this behaves like a fresh apply") in the first workflow, which I feel would be more common.

(fwiw I don't think my other argument re: users relying on it for miscellaneous reasons applies here, because for one thing they are actively opting in, and for another, invalid data is at best useless and at worst harmful.)

alexzielenski replied:
>  Can you think of any reason it would be valuable to retain in that case?

One consideration is that in the current form if an applier made use of the the `kubectl` field manager later, the annotation would be updated once again. 

> Then step 3 will repair the annotation if it is still present? 

Step 2 would have caused the annotation to be dropped since it would have no owner. It would be gone by #3.

KnVerey replied:
> One consideration is that in the current form if an applier made use of the the kubectl field manager later, the annotation would be updated once again.

If they really want that, would it work for them to add the annotation in with a blank value themselves? Experimentally, it seems to (create an object with SSA and then add `kubectl.kubernetes.io/last-applied-configuration: "{}"` to it--suddenly it will be populated!)

> Step 2 would have caused the annotation to be dropped since it would have no owner. It would be gone by no. 3.

If we don't opt it out of our migration process here, then the special field manager would be keeping it around in 2 and 3, no?

alexzielenski replied:
Considering the above, agreed it is best to not migrate the annotation if custom field manager is detected. Since its value would always be wrong anyway.

alexzielenski replied:
> If we don't opt it out of our migration process here, then the special field manager would be keeping it around in 2 and 3, no?

when a custom field manager is used:

if you make the change you propose and DO opt it out of the migration process: step 2 would drop annotation
if we do nothing, and the annotation is applied into a new separate field manager on behalf of the user: both step 2 and step 3 retain the annotation

KnVerey replied:
Sounds good. Can we also add a test (or maybe just augment the existing non-default manager test) to show we're removing the unmaintained data on purpose?

alexzielenski replied:
logic has been added and test augmented

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +697,170 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+			}
+
+			// Send the patch to upgrade the managed fields if it is non-nil
+			obj, err = helper.Patch(
```
read my mind. Just was working on a change to do this as a single PATCH call rather than the complicated loop logic. 

But I would expect this case to be pretty rare anyway.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -542,6 +547,38 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+			// This migration will be re-attempted if necessary upon next
+			// apply.
+			fmt.Fprintf(o.ErrOut, warningMigrationPatchFailed, err.Error())
+		} else if obj, err = helper.Patch(
```
1. will do
2. I think there is no need to retry the server-side-apply patch since there is no reason to expect it would conflict. @apelisse can confirm

alexzielenski replied:
fixed this bug and also added unit test logic for this case

alexzielenski committed:
add OWNERS to csaupgrade

alexzielenski committed:
add UpgradeManagedFieldsPatch

rather than modify the object directly, this function provides a JSONPATCH that should be sent to the server to upgrade its managed fields.

KnVerey commented on staging/src/k8s.io/cli-runtime/pkg/resource/helper.go:
```diff
@@ -190,7 +190,7 @@ func (m *Helper) DeleteWithOptions(namespace, name string, options *metav1.Delet
 		options = &metav1.DeleteOptions{}
 	}
-	if m.ServerDryRun {
+	if m.ServerDryRun && len(options.DryRun) == 0 {
```
I was wondering why these needed to change and found your commit message:
> helper currently has mandatory fields. this change makes it so it only overrides the request options if the request options did not request this

So, what this changeset does is change which of the helper struct's values and the options passed in takes precedence. I looked around to see if any existing callers seem to have expectations around that. Here's an example:

```go
	deleteResponse, err := resource.
		NewHelper(info.Client, info.Mapping).
		DryRun(o.DryRunStrategy == cmdutil.DryRunServer).
		DeleteWithOptions(info.Namespace, info.Name, deleteOptions)
```

In this case, I think the caller is expecting its explicit `.DryRun` to take precedence. Would you agree? I assume the place we needed this change for is our `.Patch` calls. Could we instead do the `helper.WithFieldManager(fieldManager).Patch()` pattern?

alexzielenski replied:
thanks for calling this out. Unfortunately the `WithFieldManager` function is not functional as it appears, it imperatively sets it, so would need to reset the field manager back after the operation. 

You are right it is risky to change the semantics out from under other usages. Open to suggestions. My main goal was to reuse the existing `Apply` helper to supply a field manager to `Patch` without creating too much code smell.

alexzielenski replied:
I've removed the problematic commit, and am now copying the helper object to set the field manager at point-of-use.

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -660,6 +699,177 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		info.Name,
+		types.ApplyPatchType,
+		modified,
+		nil,
```
I know a conflict here should be really, really unlikely (someone else would have had to make a change with CSA between when we got the object and this request, I think?) but should we respect the `--force-conflicts` flag? Or are we thinking that would be so unexpected that it is better to error out and let them try again? I could go either way.

alexzielenski replied:
Right, a conflict could only happen if someone else tried to modify the annotation to a different value during our RMW operation. 

I think in this case, we may want to actually enable ForceConflicts to unconditionally take ownership here. Especially since we are limiting the users who keep the annotation to `kubectl`-fieldmanagers, the annotation will be immediately updated by the server in the next patch after this is done.

wdyt @apelisse

alexzielenski replied:
talked about this offline with @apelisse a bit. The semantics about this are hard to reason about so I think the current behavior which shows migration failure warning upon conflict is safest.

KnVerey replied:
Ok, works for me!

alexzielenski committed:
add kubectl server-side apply migrate managedfields

in discussion with SIG, there is a strong interest in keeping the last-applied-configuration around for a bit longer as other tools transition for of it. This is OK since SSA maintains the annotation on kubectl's behalf on the server-side if it exists

migrate client-side-apply fields to SSA when --serverside-side is used

https://github.com/kubernetes/kubernetes/issues/107980

https://github.com/kubernetes/kubernetes/issues/108081

https://github.com/kubernetes/kubernetes/issues/107417

https://github.com/kubernetes/kubernetes/issues/112826

add test to make sure only one apply is needed after migration

KnVerey commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,160 @@ run_kubectl_server_side_apply_tests() {
+__EOF__
+  )
+
+  # Fetch the object and check to see that it does not have a field 'legacy'
```
```suggestion
  # Fetch the object and check to see that it does not have a field 'newKey'
```

KnVerey commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,160 @@ run_kubectl_server_side_apply_tests() {
+
+
+  # cleanup endpoint for diff
+  # kubectl "${kube_flags[@]:?}" delete endpoint "testendpoint"
```
what's this?

alexzielenski replied:
remnant. removed

alexzielenski committed:
update vendor

alexzielenski commented:
/retest

KnVerey commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,165 @@ run_kubectl_server_side_apply_tests() {
+  kube::test::if_has_string "$output_message" "configmap/ssa-test created"
+  kube::test::get_object_assert "configmap ssa-test" "{{ .data.key }}" 'value1'
+
+  # show that after migrating to CSA with a custom field manager, the
```
```suggestion
  # show that after client-side applying with a custom field manager, the
```

KnVerey commented on test/cmd/apply.sh:
```diff
@@ -438,6 +438,165 @@ run_kubectl_server_side_apply_tests() {
+
+  # show that after migrating to SSA with a custom field manager, the 
+  # last-applied-annotation is dropped
+  ! grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")" || exit 1
```
The CI failure is a shellcheck error about this line:

```
In ./test/cmd/apply.sh line 581:
  ! grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")"
  ^-- SC2251: This ! is not on a condition and skips errexit. Use `&& exit 1` instead, or make sure $? is checked.

Did you mean: 
   grep -q kubectl.kubernetes.io/last-applied-configuration <<< "$(kubectl get configmap ssa-test -o yaml "${kube_flags[@]:?}")" && exit 1

For more information:
  https://www.shellcheck.net/wiki/SC2251 -- This ! is not on a condition and ...
```

alexzielenski replied:
yup. I think I fixed this while your review was pending. the logic in the test is fixed

KnVerey reviewed (approved):
I have a couple more very minor suggestions, but no concerns. LGTM! 

/lgtm
/approve

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -542,6 +547,40 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+		//
+		// To check beforehand means 2 requests for most operations, and 3
+		// requests in worst case.
+		if err = o.saveLastApplyAnnotationIfNecessary(helper, info); err != nil {
+			fmt.Fprintf(o.ErrOut, warningMigrationLastAppliedFailed, err.Error())
+		} else if performedMigration, err := o.migrateToSSAIfNecessary(helper, info); err != nil {
```
Should we do the `saveLastApplyAnnotationIfNecessary` if `performedMigration` is false?

alexzielenski replied:
I dont think you should save the annotation after migrating. Ordering it that way, if preserving the annotation fails then next time user invokes server-side-apply the annotation would be dropped.

apelisse reviewed (commented):
/approve

I can lgtm once Katrina's comments have been addressed, thanks

alexzielenski commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -18,17 +18,45 @@ package csaupgrade
+// the fields in the given set
+//
+// If there is an error decoding one of the fieldsets for any reason, it is ignored
+// and assumed not to match the query.
```
todo was added here: 

https://github.com/kubernetes/kubernetes/blob/26a6e1234869b5c546195aaf416f3424cd3c3dc8/staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go#L243-L245

alexzielenski commented:
/retest

apelisse reviewed (commented):
/lgtm

lavalamp commented on staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go:
```diff
@@ -48,21 +76,107 @@ var csaAnnotationFieldSet = fieldpath.NewSet(fieldpath.MakePathOrDie("metadata",
+			"value": filteredManagers,
+		},
+		{
+			// Use "replace" instead of "test" operation so that etcd rejects with
```
We don't write this field to etcd so I don't think this explanation is right?

apelisse replied:
So if we do a test of resourceVersion, the patch fails to apply without much details, if we set the resourceVersion, then the compare-and-switch etcd performs to write the object fails, and we get a proper 409 conflict. The comment reads fine to me.

lavalamp replied:
I think it's the "etcd rejects" part that threw me off, I thought it was implying that etcd was checking the field instead of this being passed as metadata on the request.

Maybe open an issue about improving error messages from test directives in patches?

apelisse replied:
We could improve the error messages, but I don't know if the apiserver will ever interpret/return a failure to apply the patch as a 409 Conflict?

lavalamp commented:
/approve

for client-go

BenTheElder reviewed (commented):
/approve
for vendor/modules.txt

alexzielenski commented:
/hold cancel

alexzielenski commented:
/retest

fjolublar commented:
Hi @alexzielenski,
Are there any plans to bring this fix to `1.25`?
